### PR TITLE
[mn5vnp] Port common package to windows

### PR DESCRIPTION
**Pull request by Sckorn**

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ########## general ###########
 
 build --dynamic_mode=fully
-build --force_pic
+#build --force_pic
```
Comment (radameon) [inline .bazelrc:4]: Remove commented code

File: .bazelrc [hunk 2]
```diff
@@ -34,20 +34,81 @@ test --test_tag_filters=-integration_tests
...
+build:windows --host_copt=/std:c++20
+build:windows --copt=/W0
+build:windows --host_copt=/W0
+
+# On Windows, `__cplusplus` is wrongly defined without this switch
+# See https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/
+build:windows --copt=/Zc:__cplusplus
+build:windows --host_copt=/Zc:__cplusplus
+
+# Tensorflow uses M_* math constants that only get defined by MSVC headers if
```
Comment (radameon) [inline .bazelrc:55]: I believe you do not mantain TF, remove it's mention.

File: .bazelrc [hunk 3]
```diff
@@ -34,20 +34,81 @@ test --test_tag_filters=-integration_tests
...
+# _USE_MATH_DEFINES is defined.
+build:windows --copt=/D_USE_MATH_DEFINES
+build:windows --host_copt=/D_USE_MATH_DEFINES
+
+# Windows has a relatively short command line limit, which TF has begun to hit.
```
Comment (radameon) [inline .bazelrc:60]: Same about TF.

File: .bazelrc [hunk 4]
```diff
@@ -34,20 +34,81 @@ test --test_tag_filters=-integration_tests
...
+# See https://docs.bazel.build/versions/main/windows.html
+build:windows --features=compiler_param_file
+build:windows --features=archive_param_file
+
+# Speed Windows compile times. Available in VS 16.4 (we are on 16.11). See
```
Comment (radameon) [inline .bazelrc:65]: Check your MSVC, and fix comment accordingly.

File: .bazelrc [hunk 5]
```diff
@@ -34,20 +34,81 @@ test --test_tag_filters=-integration_tests
...
+# build_pip_package script in its current form (as of Aug 2023) uses the
+# runfiles symlink tree to decide what to put into the Python wheel.
+startup --windows_enable_symlinks
+build:windows --enable_runfiles
+
+# Make sure to include as little of windows.h as possible
+build:windows --copt=-DWIN32_LEAN_AND_MEAN
+build:windows --host_copt=-DWIN32_LEAN_AND_MEAN
+build:windows --copt=-DNOGDI
+build:windows --host_copt=-DNOGDI
```
Comment (radameon) [inline .bazelrc:81]: Offtop: ah, the good ol' stuff.

File: .bazelversion [hunk 6]
```diff
@@ -0,0 +1 @@
+7.0.0
```
Comment (radameon) [inline .bazelversion:1]: Bump to 7.1.2, if it is the version everywhere.

File: WORKSPACE [hunk 7]
```diff
@@ -19,30 +19,30 @@ load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")
 bazel_skylib_workspace()
 # External deps
 
-git_repository(
+#git_repository(
+#    name = "com_github_nelhage_rules_boost",
+#    commit = "fc556bef57fad1b522ddae919d978597351d5147",
+#    remote = "https://github.com/nelhage/rules_boost",
+#)
```
Comment (radameon) [inline WORKSPACE:26]: Remove commented code.

File: WORKSPACE [hunk 8]
```diff
@@ -19,30 +19,30 @@ load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")
...
+http_archive(
     name = "com_github_nelhage_rules_boost",
-    commit = "c3fae06e819ed8b93e31b150387dce4864758643",
-    remote = "https://github.com/nelhage/rules_boost",
+
+    # Replace the commit hash in both places (below) with the latest, rather than using the stale one here.
+    # Even better, set up Renovate and let it do the work for you (see "Suggestion: Updates" in the README).
+    url = "https://github.com/nelhage/rules_boost/archive/5d04542e52164931841d06d5a6b3fd2f43c436d0.tar.gz",
+    strip_prefix = "rules_boost-5d04542e52164931841d06d5a6b3fd2f43c436d0",
+    # When you first run this tool, it'll recommend a sha256 hash to put here with a message like: "DEBUG: Rule 'com_github_nelhage_rules_boost' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = ..."
```
Comment (radameon) [inline WORKSPACE:35]: Add sha256 to the call.

File: WORKSPACE [hunk 9]
```diff
@@ -19,30 +19,30 @@ load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")
...
-
-http_archive(
-    name = "gtest",
-    build_file = "@//thirdparty:gtest.BUILD",
-    sha256 = "b58cb7547a28b2c718d1e38aee18a3659c9e3ff52440297e965f5edffe34b6d0",
-    strip_prefix = "googletest-release-1.7.0",
-    url = "https://github.com/google/googletest/archive/release-1.7.0.zip",
+  name = "com_google_googletest",
+  urls = ["https://github.com/google/googletest/archive/5ab508a01f9eb089207ee87fd547d290da39d015.zip"],
+  strip_prefix = "googletest-5ab508a01f9eb089207ee87fd547d290da39d015",
```
Comment (radameon) [inline WORKSPACE:45]: sha256

File: common/BUILD [hunk 10]
```diff
@@ -1,38 +1,48 @@
...
-    ]),
+    ],
     copts = ["-Iexternal/nlohmann/include"],
     include_prefix = "rcbe-engine",
     includes = ["include/common"],
-    linkopts = ["-lpthread"],
+    linkopts = select({
+        "@bazel_tools//src/conditions:windows": [""],
+        "//conditions:default": ["-lpthread"],
+    }),
```
Comment (radameon) [inline common/BUILD:31]: Try to do it up the BUILD file, and just use the private var here.

File: common/BUILD [hunk 11]
```diff
@@ -1,38 +1,48 @@
...
+    deps = ["@nlohmann//:nlohmann_json"],
 )
 
 cc_library(
     name = "engine_fundamentals",
     hdrs = glob([
         "include/common/fundamentals/**/*.hpp",
     ]),
+    data = select({
+        "@bazel_tools//src/conditions:windows": [],
```
Comment (radameon) [inline common/BUILD:43]: Why not?

File: common/BUILD [hunk 12]
```diff
@@ -1,38 +1,48 @@
...
-    deps = ["@nlohmann"],
+    deps = ["@nlohmann//:nlohmann_json"],
 )
 
 cc_library(
     name = "engine_fundamentals",
     hdrs = glob([
         "include/common/fundamentals/**/*.hpp",
     ]),
+    data = select({
```
Comment (radameon) [inline common/BUILD:42]: Moreover, why data?

File: common/BUILD [hunk 13]
```diff
@@ -123,8 +133,9 @@ cc_test(
     deps = [
         ":rcbe_binary_tools",
         ":rcbe_delegate",
-        "@gtest",
+        "@com_google_googletest//:gtest_main",
     ],
+    linkstatic = 1,
```
Comment (radameon) [inline common/BUILD:138]: Is it really needed? If so, should depend on flag?

File: common/include/common/binary/BinaryBuffer.hpp [hunk 14]
```diff
@@ -112,44 +112,44 @@ class BinaryBuffer {
...
-    void appendImplementation(const StorageType &storage);
-    void appendImplementation(const ViewType &storage);
+    R_PUBLIC_API BinaryBuffer &appendImplementation(const BinaryBuffer &source);
+    R_PUBLIC_API void appendImplementation(const StorageType &storage);
+    R_PUBLIC_API void appendImplementation(const ViewType &storage);
 
-    std::vector<ByteType> buffer_;
-    std::vector<const ByteType *> buffer_view_;
+    std::vector<ByteType> buffer_; // TODO: StorageType buffer_;
+    std::vector<const ByteType *> buffer_view_; // TODO: ViewType buffer_view_;
```
Comment (radameon) [inline common/include/common/binary/BinaryBuffer.hpp:120]: Replace the types with aliases, if you're up to it anyway.

File: common/test/binary_file_tests.cpp [hunk 15]
```diff
@@ -67,7 +67,12 @@ class BinaryFileTests : public ::testing::Test {
 
 private:
 
+/// TODO: introduce wrapper for different OS paths, or better use std::filesystem @radameon @sckorn
+#ifdef __linux__
     const std::string file_path_str_ = "/tmp/binary_test.dat";
+#elif _WIN32
+    const std::string file_path_str_ = "C:\\Windows\\Temp\\binary_test.dat";
```
Comment (radameon) [inline common/test/binary_file_tests.cpp:74]: At lease try to fix it now.

File: meta_package/windows_api/public_api.hpp [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+#define R_PUBLIC_API __declspec(dllexport)
+
+#else
+
+#define R_PUBLIC_API __declspec(dllimport)
+
+#endif
+
+#elif defined(__linux__)
```
Comment (radameon) [inline meta_package/windows_api/public_api.hpp:16]: ANd what if the flag is set to static libs?

**Review by radameon — CHANGES_REQUESTED** [review]
Needs some fixing

File: .bazelversion (hunk 6 above)
Comment (Sckorn) [reply .bazelversion:1]: Nope. 7.0.0 inside docker, gotta keep it like that for now.

File: .bazelrc [hunk 17]
```diff
@@ -34,20 +33,81 @@ test --test_tag_filters=-integration_tests
...
+
+# Speed Windows compile times. Available in VS 16.4 (we are on 17.0.1). See
+# https://groups.google.com/a/tensorflow.org/d/topic/build/SsW98Eo7l3o/discussion
+build:windows --copt=/d2ReducedOptimizeHugeFunctions
+build:windows --host_copt=/d2ReducedOptimizeHugeFunctions
+
+# Enable the runfiles symlink tree on Windows. This makes it possible to build
+# the pip package on Windows without an intermediate data-file archive, as the
+# build_pip_package script in its current form (as of Aug 2023) uses the
+# runfiles symlink tree to decide what to put into the Python wheel.
```
Comment (radameon) [inline .bazelrc:72]: This comment, at least now, is not relevant for our repo. Remove it.

**Review by radameon — CHANGES_REQUESTED** [review]
Fix one comment, and we're ready

**Review by radameon — APPROVED** [review]
LGTM

**Pull request merged**